pika/adapters/base_connection.py

```diff
@@ -347,23 +356,43 @@ def _handle_read(self):
             LOGGER.error('Read empty data, calling disconnect')
             return self._handle_disconnect()
 
+        #LOGGER.debug("read %s bytes", len(data))
         # Pass the data into our top level frame dispatching method
         self._on_data_available(data)
         return len(data)
 
     def _handle_write(self):
-        """Handle any outbound buffer writes that need to take place."""
+        """Try and write as much as we can, if we get blocked requeue 
+        what's left"""
         bytes_written = 0
-        if self.outbound_buffer:
-            frame = self.outbound_buffer.popleft()
-            try:
-                self.socket.sendall(frame)
-                bytes_written = len(frame)
-            except socket.timeout:
-                raise
-            except socket.error as error:
+        try:
+            while self.outbound_buffer:
+                frame = self.outbound_buffer.popleft()
+                bw = self.socket.send(frame)
+                frame = frame[bw:]
+                bytes_written += bw
+
+                if frame:
+                    LOGGER.warning("Partial write, requeing remaining data")
+                    self.outbound_buffer.appendleft(frame)
+                    break
+
+        except socket.timeout:
+            # Will only come here if the socket is blocking
+            LOGGER.warning("socket timeout, requeuing frame")
+            self.outbound_buffer.appendleft(frame)
```

Comment:
If `socket.timeout` exception is going to be suppressed in `_handle_write()`, it necessitates clean-up in `BlockingConnection._flush_outbound()`, which still expects to handle `socket.timeout` from `BaseConnection._handle_write()`, which becomes dead code with this change. Then there are also `BlockingConnection._handle_timeout()` and `BlockingConnection._socket_timeouts` logic bits that need to be considered.